pkg/bzz/address.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package bzz
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+
+	"github.com/ethersphere/bee/pkg/crypto"
+	"github.com/ethersphere/bee/pkg/swarm"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+var ErrInvalidAddress = errors.New("invalid address")
+
+type Address struct {
+	Underlay  ma.Multiaddr
+	Overlay   swarm.Address
+	Signature []byte
+}
+
+type addressJSON struct {
+	Overlay   string
+	Underlay  string
+	Signature string
+}
+
+func NewAddress(signer crypto.Signer, underlay ma.Multiaddr, overlay swarm.Address, networkID uint64) (*Address, error) {
+	networkIDBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(networkIDBytes, networkID)
+	signature, err := signer.Sign(append(underlay.Bytes(), networkIDBytes...))
+	if err != nil {
+		return nil, err
+	}
+
+	return &Address{
+		Underlay:  underlay,
+		Overlay:   overlay,
+		Signature: signature,
+	}, nil
+}
+
+func Parse(underlay, overlay, signature []byte, networkID uint64) (*Address, error) {
+	networkIDBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(networkIDBytes, networkID)
+	recoveredPK, err := crypto.Recover(signature, append(underlay, networkIDBytes...))
+	if err != nil {
+		return nil, ErrInvalidAddress
+	}
+
+	recoveredOverlay := crypto.NewOverlayAddress(*recoveredPK, networkID)
+	if !bytes.Equal(recoveredOverlay.Bytes(), overlay) {
+		return nil, ErrInvalidAddress
+	}
+
+	multiUnderlay, err := ma.NewMultiaddrBytes(underlay)
+	if err != nil {
+		return nil, ErrInvalidAddress
+	}
+
+	return &Address{
+		Underlay:  multiUnderlay,
+		Overlay:   swarm.NewAddress(overlay),
+		Signature: signature,
+	}, nil
+}
+
+func (a *Address) Equal(b *Address) bool {
+	return a.Overlay.Equal(b.Overlay) && a.Underlay.Equal(b.Underlay) && bytes.Equal(a.Signature, b.Signature)
+}
+
+func (p *Address) MarshalJSON() ([]byte, error) {
+	return json.Marshal(&addressJSON{
+		Overlay:   p.Overlay.String(),
+		Underlay:  p.Underlay.String(),
+		Signature: base64.StdEncoding.EncodeToString(p.Signature),
+	})
+}
+
+func (p *Address) UnmarshalJSON(b []byte) error {
+	v := &addressJSON{}
+	err := json.Unmarshal(b, v)
+	if err != nil {
+		return err
+	}
+
+	a, err := swarm.ParseHexAddress(v.Overlay)
+	if err != nil {
+		return err
+	}
+
+	p.Overlay = a
+
+	m, err := ma.NewMultiaddr(v.Underlay)
+	if err != nil {
+		return err
+	}
+
+	p.Underlay = m
+	p.Signature, err = base64.StdEncoding.DecodeString(v.Signature)
```

Comment:
i would like to raise a flag here: you're not verifying that the signature is actually correct here. so theoretically i could somehow persist a wrong signature, deserialize it and it would turn out as a valid address, as opposed to `Parse` or `New` that actually calculate the signature. I would consider checking it here too

Reply:
The reasons why I did not add this check here is
1. It would need a networkID to be provided as well for the check
2. I somehow see it similarly to `&BzzAddress{ ... }` cuz it is just marshalling, unmarhsalling json. 

I can add the networkID to the function parameters as well, but I am not sure this is much useful, as these functions are made for  purpose of persisting  in the addressbook.

Reply:
> My opinion is, if we end up using this on many places and need the check, then I would make these functions with some factory-style receivers or something similar.

I agree.